### PR TITLE
[RW-10031][risk=no] Periodically refresh the Leo cookie

### PR DESCRIPTION
**Pull request by Peter-Lavigne**
Periodically refreshes the Leo cookie to match Terra's implementation. This removes our reliance on the [Jupyter auth extension](https://github.com/DataBiosphere/terra-docker/blob/c762c950896983e760015460ab9a29ab9411e249/terra-jupyter-base/custom/google_sign_in.js).

[This thread](https://broadinstitute.slack.com/archives/CA3NP1733/p1683126985418719) provides additional background info.

Unfortunately, this does not fully resolve the problem because there is still contention between the extension and our implementation. To completely fix the bug we will need to remove our reliance on the extension. After merging this, I will find a way to remove the contention with the extension (Terra is not impacted by this contention but I haven't investigated why yet) or deprecate the extension entirely.

To manually test this, load any page and observe the `setCookie` endpoint being called every five minutes.

We could use `withLeoCookie` only on the pages that actually need it. However, I think converging on their approach is acceptable to alleviate the "it works in Terra" problem we're running into with auth.

[Relevant #workbench-ux thread](https://pmi-engteam.slack.com/archives/C02MWP2RN5P/p1683665947528789).

**Manually testing the AbortController**

Testing the abort behavior with an automated test is difficult for reasons described [here](https://github.com/all-of-us/workbench/pull/7664). Instead, I have tested this manually.

See my testing process described [here](https://github.com/all-of-us/workbench/pull/7664). However, in order to unmount the component before the page loads, I added `<Link to='/not-found'>TEST UNMOUNT ABORT</Link>` immediately below the `<AppRouter>` line in `ui/src/app/routing/app-routing.tsx`. Clicking on that while the request is pending will trigger an abort.

**Screenshots**

Existing behavior: When `setCookie` fails for Jupyter:
image

New behavior: When `setCookie` fails (not app-specific):
image (1)

---
**PR checklist**

- [x] I have included an issue ID or "no ticket" in the PR title as outlined in [CONTRIBUTING.md](https://github.com/all-of-us/workbench/blob/main/.github/CONTRIBUTING.md).
- [x] I have included a risk tag of the form `[risk=no|low|moderate|severe]` in the PR title as outlined in [CONTRIBUTING.md](https://github.com/all-of-us/workbench/blob/main/.github/CONTRIBUTING.md).
- [x] I have manually tested this change and my testing process is described above.
- [x] This change includes appropriate automated tests, and I have documented any behavior that cannot be tested with code.
- [x] I have added explanatory comments where the logic is not obvious.
- One or more of the following is true:
  - [ ] This change is intended to complete a JIRA story, so I have checked that all AC are met for that story.
  - [ ] This change fixes a bug, so I have ensured the steps to reproduce are in the Jira ticket or provided above.
  - [ ] This change impacts deployment safety (e.g. removing/altering APIs which are in use), so I have documented the impacts in the description.
  - [ ] This change includes a new feature flag, so I have created and linked new JIRA tickets to (a) turn on the feature flag and (b) remove it later.
  - [x] This change modifies the UI, so I have taken screenshots or recordings of the new behavior and notified the PO and UX designer in [Slack](https://pmi-engteam.slack.com/archives/C02MWP2RN5P).
  - [ ] This change modifies API behavior, so I have run the relevant E2E tests locally because API changes are not covered by our PR checks.
  - [ ] None of the above apply to this change.

### CODE REVIEW

File: ui/src/app/pages/analysis/leonardo-app-launcher.tsx [hunk 1]
```diff
@@ -460,17 +460,6 @@ export const LeonardoAppLauncher = fp.flow(
       return appendNotebookFileSuffix(this.getNotebookName());
     }
 
-    private async initializeNotebookCookies() {
-      return await this.runtimeRetry(() =>
```
Comment (Peter-Lavigne) [inline ui/src/app/pages/analysis/leonardo-app-launcher.tsx:464]: Why were we retrying this, and should I keep that functionality? An error modal telling the user to refresh on the first failure may be sufficient.

Comment (Peter-Lavigne) [inline ui/src/app/pages/analysis/leonardo-app-launcher.tsx:464]: Should we keep the retry functionality? I think an error modal telling the user to refresh on the first failure may be sufficient.

File: ui/src/app/components/apps-panel/expanded-app.spec.tsx [hunk 2]
```diff
@@ -345,22 +350,32 @@ describe('ExpandedApp', () => {
...
       status: AppStatus.RUNNING,
+      proxyUrls: {
+        rstudio: proxyUrl,
+      },
     });
 
+    const focusStub = jest.fn();
+    const windowOpenSpy = jest
+      .spyOn(window, 'open')
+      // @ts-ignore
```
Comment (jmthibault79) [inline ui/src/app/components/apps-panel/expanded-app.spec.tsx:366]: ?

File: ui/src/app/components/apps-panel/expanded-app.spec.tsx [hunk 3]
```diff
@@ -81,6 +80,12 @@ describe('ExpandedApp', () => {
       runtime: runtimeStub.runtime,
       runtimeLoaded: true,
     });
+
```
Comment (jmthibault79) [inline ui/src/app/components/apps-panel/expanded-app.spec.tsx:83]: the changes in this file do not seem to be related?

File: ui/src/app/utils/authentication.tsx [hunk 4]
```diff
@@ -71,6 +71,8 @@ export const makeOIDC = (config: ConfigResponse): AuthProviderProps => {
     onSigninCallback: (_user: User | void): void => {
       window.history.replaceState({}, document.title, window.location.pathname);
     },
+    // https://github.com/DataBiosphere/terra-ui/blob/356f27342ff44d322b2b52077fa4efb1c5f920ce/src/libs/auth.js#LL49C10-L49C10
```
[the remark this answers is not in the document]
Comment (jmthibault79) [reply ui/src/app/utils/authentication.tsx:74]: [after chat] Nope, that requirement is not correct.

How about: defines the access token refresh interval to be much shorter than the default of 60 min

File: ui/src/app/components/with-leo-cookie.tsx [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+          console.log('Done setting cookie.');
+          setLoading(false);
+        });
+    };
+
+    setLeoCookie();
+    const timer = setInterval(
+      setLeoCookie,
+      // https://github.com/DataBiosphere/terra-ui/blob/356f27342ff44d322b2b52077fa4efb1c5f920ce/src/libs/auth.js#LL49C10-L49C10
+      (auth.settings.accessTokenExpiringNotificationTimeInSeconds - 30) * 1000
```
[the remark this answers is not in the document]
Comment (jmthibault79) [reply ui/src/app/components/with-leo-cookie.tsx:46]: [after chat w/ new understanding]

I understand why the other value is referenced now: the cookie interval must be shorter than the auth refresh interval, and also shorter than [token default expiration - auth refresh time].  Could you indicate that in the comment?

**Review by jmthibault79 — APPROVED** [review]
Can confirm that setCookie was called many times when I left the app running.

Some concerns with some RStudio code which doesn't seem to be related, and I did not do a meaningful review of the e2ev2 code.

File: ui/src/app/utils/authentication.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (Peter-Lavigne) [reply ui/src/app/utils/authentication.tsx:74]: I've updated it to "This setting reduces the token refresh interval to 54.5 minutes (= 60 minutes - 330 seconds)."

File: ui/src/app/components/with-leo-cookie.tsx [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+const LeoCookieWrapper = ({ children }) => {
+  const { auth } = useStore(authStore);
+
+  const [loading, setLoading] = useState(true);
+
+  useEffect(() => {
+    const pollAborter = new AbortController();
+
+    const setLeoCookie = () => {
+      console.log('setting cookie');
```
[the remark this answers is not in the document]
Comment (Peter-Lavigne) [reply ui/src/app/components/with-leo-cookie.tsx:16]: Unintentional, removed.

File: ui/src/app/utils/authentication.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (jmthibault79) [reply ui/src/app/utils/authentication.tsx:74]: oh!  thank you.  that is not what I understood, so this clarification is clearly needed :)

File: ui/src/app/components/with-leo-cookie.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (Peter-Lavigne) [reply ui/src/app/components/with-leo-cookie.tsx:46]: I've added several lines of comments to explain why we do this. The link to Terra's comment doesn't seem sufficient unless you already understand the situation.

File: ui/src/app/components/with-leo-cookie.tsx [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+    setLeoCookie();
+    const timer = setInterval(
+      setLeoCookie,
+      // https://github.com/DataBiosphere/terra-ui/blob/356f27342ff44d322b2b52077fa4efb1c5f920ce/src/libs/auth.js#LL49C10-L49C10
+      (auth.settings.accessTokenExpiringNotificationTimeInSeconds - 30) * 1000
+    );
+
+    return () => {
+      clearTimeout(timer);
+      console.log('aborting');
```
[the remark this answers is not in the document]
Comment (Peter-Lavigne) [reply ui/src/app/components/with-leo-cookie.tsx:51]: Also unintentional, removed.

It might be worth adding a linter at some point that comments on PRs: "Is this log statement intentional"? 😄

File: ui/src/app/components/with-leo-cookie.spec.tsx [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+    jest.useFakeTimers();
+
+    const leoProxyApiStub = new LeoProxyApiStub();
+    leoRegisterApiClient(ProxyApi, leoProxyApiStub);
+
+    authStore.set({
+      authLoaded: true,
+      isSignedIn: true,
+      auth: {
+        // @ts-ignore
```
[the remark this answers is not in the document]
Comment (Peter-Lavigne) [reply ui/src/app/components/with-leo-cookie.spec.tsx:32]: Good callout 😄 

Primarily, 10 other irrelevant fields:

image

Other options include:
- Creating a valid mock object and overriding this value
- Depending on only `accessTokenExpiringNotificationTimeInSeconds` in the component, somehow

File: ui/src/app/components/with-leo-cookie.spec.tsx [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+        settings: {
+          accessTokenExpiringNotificationTimeInSeconds: 500,
+        },
+      },
+    });
+    notificationStore.set(null);
+  });
+
+  it('should render the children once the cookie loads', async () => {
```
[the remark this answers is not in the document]
Comment (Peter-Lavigne) [reply ui/src/app/components/with-leo-cookie.spec.tsx:41]: `with-leo-cookie` _does_ show the child component if there's an error, actually. Otherwise, after clicking out of the modal, there will be a completely blank screen:

image

I think this is a limitation of our current top-level error handling. It would be nice to see an error page if calls to config, profile, etc. fail.

[the remark this answers is not in the document]
Comment (Peter-Lavigne) [reply ui/src/app/components/with-leo-cookie.spec.tsx:41]: I'm open to changing this. However, as is, I think having three tests is OK because rendering children and showing errors are unrelated

File: ui/src/app/components/apps-panel/expanded-app.spec.tsx (hunk 2 above)
Comment (Peter-Lavigne) [reply ui/src/app/components/apps-panel/expanded-app.spec.tsx:366]: Similar to the above but more extreme:

image

I'm not too concerned about perfecting this particular test since [RW-9664](https://precisionmedicineinitiative.atlassian.net/browse/RW-9664) will remove the need for it entirely.

File: ui/src/app/components/with-leo-cookie.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (jmthibault79) [reply ui/src/app/components/with-leo-cookie.tsx:50]: very nice, thanks

File: ui/src/app/components/with-leo-cookie.spec.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (jmthibault79) [reply ui/src/app/components/with-leo-cookie.spec.tsx:32]: ok, this is fine

File: ui/src/app/components/apps-panel/expanded-app.spec.tsx (hunk 2 above)
Comment (jmthibault79) [reply ui/src/app/components/apps-panel/expanded-app.spec.tsx:366]: This works but I wouldn't claim that it's better
```
.mockReturnValue({ focus: focusStub } as any as Window);
```

Comment (Peter-Lavigne) [reply ui/src/app/components/apps-panel/expanded-app.spec.tsx]: I like that better since I think the intent is clearer. Updated.

**Comment by Peter-Lavigne** [comment]
Each failing E2E test has passed on at least one run. Additionally, the tests seem unrelated to the change. I think it's OK to merge this PR with these tests failing.
